openssl/src/ts.rs

```diff
@@ -0,0 +1,457 @@
+        /// Serializes the timestamp request into a DER-encoded TimeStampResp structure.
+        ///
+        /// This corresponds to [`i2d_TS_RESP`].
+        ///
+        /// [`i2d_TS_RESP`]: https://www.openssl.org/docs/man1.1.0/man3/i2d_TS_RESP.html
+        to_der,
+        ffi::i2d_TS_RESP
+    }
+
+    /// Verifies a timestamp response.
+    ///
+    /// This corresponds to `TS_RESP_verify_response`.
+    pub fn verify(&self, context: &TsVerifyContext) -> Result<(), ErrorStack> {
+        unsafe {
+            cvt(ffi::TS_RESP_verify_response(
+                context.as_ptr(),
+                self.as_ptr(),
+            ))
+            .map(|_| ())
+        }
+    }
+}
+
+bitflags! {
+    /// Flags controlling timestamp verification behaviour.
+    pub struct VerifyFlags: c_uint {
+        const SIGNATURE = ffi::TS_VFY_SIGNATURE;
+        const VERSION = ffi::TS_VFY_VERSION;
+        const POLICY = ffi::TS_VFY_POLICY;
+        const IMPRINT = ffi::TS_VFY_IMPRINT;
+        const DATA = ffi::TS_VFY_DATA;
+        const NONCE = ffi::TS_VFY_NONCE;
+        const SIGNER = ffi::TS_VFY_SIGNER;
+        const TSA_NAME = ffi::TS_VFY_TSA_NAME;
+
+        const ALL_IMPRINT = ffi::TS_VFY_ALL_IMPRINT;
+        const ALL_DATA = ffi::TS_VFY_ALL_DATA;
+    }
+}
+
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::TS_VERIFY_CTX;
+    fn drop = ffi::TS_VERIFY_CTX_free;
+
+    /// A context object specifying time-stamping response verification parameters.
+    pub struct TsVerifyContext;
+
+    /// Reference to `TsVerifyContext`.
+    pub struct TsVerifyContextRef;
+}
+
+impl TsVerifyContext {
+    /// Construct a verify context from a timestamping request.
+    ///
+    /// Corresponds to `TS_REQ_to_TS_VERIFY_CTX`.
+    pub fn from_req(request: &TsReqRef) -> Result<TsVerifyContext, ErrorStack> {
+        unsafe {
+            ffi::init();
+            let ctx: *mut ffi::TS_VERIFY_CTX = cvt_p(ffi::TS_VERIFY_CTX_new())?;
+            let ctx = cvt_p(ffi::TS_REQ_to_TS_VERIFY_CTX(request.as_ptr(), ctx))?;
+            Ok(TsVerifyContext::from_ptr(ctx))
+        }
+    }
+}
+
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::TS_RESP_CTX;
+    fn drop = ffi::TS_RESP_CTX_free;
+
+    /// A context object used to sign timestamp requests.
+    pub struct TsRespContext;
+
+    /// Reference to `TsRespContext`.
+    pub struct TsRespContextRef;
+}
+
+impl TsRespContextRef {
+    /// Creates a signed timestamp response for the request.
+    ///
+    /// This corresponds to `TS_RESP_create_response`.
+    pub fn create_response(&mut self, request: &TsReqRef) -> Result<TsResp, ErrorStack> {
+        unsafe {
+            let der = request.to_der()?;
+            let bio = ::bio::MemBioSlice::new(&der)?;
+            let response = cvt_p(ffi::TS_RESP_create_response(self.as_ptr(), bio.as_ptr()))?;
+            Ok(TsResp::from_ptr(response))
+        }
+    }
+}
+
+pub struct TsRespContextBuilder(TsRespContext);
+
+impl TsRespContextBuilder {
```

Comment:
Probably same here WRT not needing a builder.

Reply:
Fixed, thanks for the review!